sceptre/resolvers/__init__.py

```diff
@@ -49,62 +49,231 @@ def resolve(self):
         self.logger = logging.getLogger(__name__)
-        self._get_in_progress = False
+
         self._lock = RLock()
 
-    def __get__(self, instance, type):
+    def __get__(self, stack: 'stack.Stack', stack_class: Type['stack.Stack']) -> Any:
         """
-        Attribute getter which resolves any Resolver object contained in the
-        complex data structure.
+        Attribute getter which resolves the resolver(s).
 
+        :param stack: The Stack instance the property is being retrieved for
+        :param stack_class: The class of the stack that the property is being retrieved for.
         :return: The attribute stored with the suffix ``name`` in the instance.
-        :rtype: dict or list
+        :rtype: The obtained value, as resolved by the property
         """
-        with self._lock, self._no_recursive_get():
-            def resolve(attr, key, value):
-                try:
-                    attr[key] = value.resolve()
-                except RecursiveGet:
-                    attr[key] = self.ResolveLater(instance, self.name, key,
-                                                  lambda: value.resolve())
-
-            if hasattr(instance, self.name):
-                retval = _call_func_on_values(
-                    resolve, getattr(instance, self.name), Resolver
-                )
-                return retval
+        with self._lock, self._no_recursive_get(stack):
+            if hasattr(stack, self.name):
+                return self.get_resolved_value(stack, stack_class)
 
-    def __set__(self, instance, value):
+    def __set__(self, stack: 'stack.Stack', value: Any):
         """
         Attribute setter which adds a stack reference to any resolvers in the
         data structure `value` and calls the setup method.
 
+        :param stack: The Stack instance the value is being set onto
+        :param value: The value being set on the property
+        """
+        with self._lock:
+            self.assign_value_to_stack(stack, value)
+
+    @contextmanager
+    def _no_recursive_get(self, stack: 'stack.Stack'):
+        # We don't care about recursive gets on the same property but different Stack instances,
+        # only recursive gets on the same stack. Some Resolvers access the same property on OTHER
+        # stacks and that actually shouldn't be a problem. Remember, these descriptor instances are
+        # set on the CLASS and so instance variables on them are shared across all classes that
+        # access them. Thus, we set this "get_in_progress" attribute on the stack instance rather
+        # than the descriptor instance.
+        get_status_name = f'_{self.name}_get_in_progress'
+        if getattr(stack, get_status_name, False):
+            raise RecursiveResolve(f"Resolving Stack.{self.name[1:]} required resolving itself")
+        setattr(stack, get_status_name, True)
+        try:
+            yield
+        finally:
+            setattr(stack, get_status_name, False)
```

Comment:
Previously, this lock would prevent you from accessing the same property on OTHER stacks. This was a bug I had to fix so that Stack A's resolver can access that same property on Stack B. This is what will happen with the StackOutput resolver when it tries to access the `iam_role` of the target stack.